### DISCUSSION TITLE
Can not include floorplan.yaml

### DISCUSSION THREAD

**Discussion by gillman99** [ExperienceLovelace/ha-floorplan#20] [Q&A] [answered]
exetico,
Thank you for your help in getting the panel working but when I copy your code into the raw editor I get an error on !include.  Here's what I have in my editor:

  - title: Floorplan
    icon: 'mdi:floor-plan'
    panel: true
    cards:
      - type: vertical-stack
        cards:
          - type: horizontal-stack
            cards:
              - config: !include /local/floorplan/floorplan.yaml
                type: 'custom:floorplan-card'
                full_height: true

Here's the error message:

<img width="595" alt="Screen Shot 2021-01-17 at 7 23 30 AM" src="https://user-images.githubusercontent.com/76920134/104844269-e8425f00-5894-11eb-8279-5ca341e2d97d.png">

Attached is my floorplan.yaml:

      name: Mark Floorplan
      image: /local/floorplan/house2.svg
      stylesheet: /local/floorplan/floorplan.css

      groups:

        - name: Lights
          entities:
             - light.caseta_r_wireless_in_wall_dimmer_2
             - light.caseta_r_wireless_in_wall_dimmer_3
             - light.caseta_r_wireless_in_wall_dimmer_4
             - light.caseta_r_wireless_in_wall_dimmer_5
          states:
            - state: 'on'
              class: 'light-on'
            - state: 'off'
              class: 'light-off'
          action:
            domain: homeassistant
            service: toggle

Any idea what's causing the error in !include?

**Comment by exetico** [comment]
Maybe due to www vs. local.

I've created a folder for all my Floorplan-files, which are called /lovelace. Could you try that?

![Example of YAML](https://user-images.githubusercontent.com/3549445/104515713-74355c00-55f3-11eb-94a9-a4225e858106.jpg)

Another think could be the use of `/` in the front, but I'm not 100% sure (I'm om my phone right now)...

**Reply by gillman99** [reply]
Exetico,
By /lovelace, should I put that in root, under config, someplace else?

**Reply by exetico** [reply]
Normally the HA-root are the folder where your configuration.yaml are saved :-) And yes, `lovelace` should be created in the same folder as your configuration.yaml-file.

```bash
--- /home/yoga/homeassistant -----------------------------------
    8.8 GiB [##########]  home-assistant_v2.db                  
    7.6 MiB [          ] /www
    4.0 MiB [          ] /custom_components
  284.0 KiB [          ] /themes
  112.0 KiB [          ] /lovelace
   68.0 KiB [          ] /packages
   44.0 KiB [          ]  ui-lovelace.yaml
   24.0 KiB [          ]  configuration.yaml
   ```

Here's an example with some of my files :-)

My YAML are located in `configuration.yaml`:
```yaml
  - title: Floorplan
    icon: 'mdi:floor-plan'
    panel: true
    cards:
      - type: vertical-stack
        cards:
          - type: horizontal-stack  
            cards:
              - config: !include lovelace/floorplan/_config-floorplan1.yaml
                type: 'custom:floorplan-card'
                full_height: true
```

And the configs are located within the lovelace-folder. 

!include are a bit hard to get use to, but you'll get there :-) 

Your floorplan1.yaml should have a formatting like this:
```yaml
image: /local/floorplan/floorplan1.svg?v=1.1.14
#log_level: error
rules: <!-- Rules goes here -->
id: floorplan1
```

But, you're asking for general HA-support. It's not really anything related to floorplan - but I know it's kinda relevant, to get a better overview of your floorplan-configs :-)

**Comment by gillman99** [comment]
One more thing Exotico, this error is happening when I'm trying to save the code in the raw configuration editor and not when I'm trying to run anything.   I think it's complaining about !include since save most likely doesn't resolve the path of what's being included.

**Reply by exetico** [reply]
Oh, fair. Maybe the include are making some validations in the GUI. which I'm not use to in the raw-editing way :-) I hope you'll get it to work, one way or another. Please share the solution here, if the above guidance are not the right answer for you.

**Comment by gillman99** [comment]
I went a different approach to get past this !include issue and am not including floorplan.yaml from my raw configuration editor yaml.  Below is a screenshot of the SVG file:

<img width="624" alt="Screen Shot 2021-01-17 at 2 12 25 PM" src="https://user-images.githubusercontent.com/76920134/104854728-09737180-58ce-11eb-8e74-8a009bb1a291.png">

Now a screenshot of the Home Assistant view:
<img width="682" alt="Screen Shot 2021-01-17 at 2 12 51 PM" src="https://user-images.githubusercontent.com/76920134/104854740-18f2ba80-58ce-11eb-9d61-5660f4931360.png">

I added some code to modify the icon of the 1 light when it is turned on or off per the demo.  Why does it size so large and not keep the same size as the light_off/on.svg file I imported into Inkscape?  Here's my yaml:

  - title: Floorplan2
    panel: true
    groups:
      - name: Lights
        entities:
          - light.caseta_r_wireless_in_wall_dimmer_2
          - light.caseta_r_wireless_in_wall_dimmer_3
          - light.caseta_r_wireless_in_wall_dimmer_4
          - light.caseta_r_wireless_in_wall_dimmer_5
        states:
          - state: 'on'
            class: light-on
          - state: 'off'
            class: light-off
        action:
          service: toggle
    cards:
      - type: 'custom:floorplan-card'
        full_height: true
        config:
          image: /local/floorplan/house2.svg?v=1.1.14
          stylesheet: /local/floorplan/floorplan.css
          defaults:
            hover_action: hover-info
          rules:
            - entity: light.caseta_r_wireless_in_wall_dimmer_5
              element: light.caseta_r_wireless_in_wall_dimmer_5
              state_action:
                action: call-service
                service: floorplan.image_set
                service_data: '/local/custom_ui/images/light_${entity.state}.svg'
              tap_action: toggle
            - entity: light.caseta_r_wireless_in_wall_dimmer_5
              element: light.caseta_r_wireless_in_wall_dimmer_5.button
              state_action:
                action: call-service
                service: floorplan.class_set
                service_data: 'button-${entity.state}'
              tap_action:
                action: call-service
                service: homeassistant.toggle

Also included my css file
[floorplan.txt](https://github.com/ExperienceLovelace/ha-floorplan/files/5826767/floorplan.txt)

Can anyone help with why my icon does not size correctly with the above code?

**Reply by exetico** [reply]
Please create another topic for that part. It's much better to have one thread for each topic (That's why I'm asking people to use the discussion here, instead of our mega-thread on the HA-community forum).

Also, you must have done something wierd here. I've not seen anything like that before. Maybe you've make an error in the SVG; or something like that? Try and use the "Inspect" option in your browser, and see why the size is changed. Maybe it due to other CSS-rules, which are conflicing. I can't find any good reason for it, so something must be a conflict or so. Are you facing the same issue, while using the sample SVG file? 

<hr>


_And oh, if you're trying to ask for help, it's much better to try and get it to work - or at least give "Thanks" in return, now that I'm spending time on helping you, the best I can, instead of just jumping to another topic_  👍  Thanks!

**Reply by gillman99** [reply]
Exetico,
I'm very appreciative of your assistance.  Been constantly working on this for the last 5 hours or so and have different options going in parallel to see if I can get one of them to work.  I'll start another thread for the sizing issue and have looked at inspect but not sure what to look for in there.  I'm a noob so learning as I go.

Again, very thankful for your assistance and guidance you've provided so far.
Mark

**Reply by exetico** [reply]
Thanks Mark. I'm sorry if that was a bit too aggressive ❤️😁

Please share the SVG, too.

**Reply by vajonam** [reply]
> Now a screenshot of the Home Assistant view:
> <img alt="Screen Shot 2021-01-17 at 2 12 51 PM" width="682" src="https://user-images.githubusercontent.com/76920134/104854740-18f2ba80-58ce-11eb-9d61-5660f4931360.png">
> 


This is most likely due the the transform that inkscape added to your imported SVG object, when set a new class to it you need to set the transform as well, or your can move and size the object and then select clear transform in inkscape.

**Comment by gillman99** [comment]
Tobias,
No problem.  How do I upload .svg file since it’s not one of the supported formats for gitHub discussions?
Mark

**Reply by exetico** [reply]
It's up to you. Just one of the major upload-services. Mega or so, is fine.

**Comment by gillman99** [comment]
Thanks.  I’ll give this a shot.

**Comment by vajonam** [comment]
AFAIK you dont need the `!include` in here just point the yaml file with the config. 

This is what I have. I am using a grid, but you can see all the configs are in each of the `mainfloor.yaml` `secondfloor.yaml` and `basment.yaml` 

```
type: vertical-stack
cards:
  - type: grid
    columns: 2
    cards:
      - type: 'custom:floorplan-card'
        full_height: false
        config: /local/floorplan/mainfloor.yaml
      - type: 'custom:floorplan-card'
        full_height: false
        config: /local/floorplan/secondfloor.yaml
      - type: 'custom:floorplan-card'
        full_height: false
        config: /local/floorplan/basement.yaml
```

**Comment by gillman99** [comment]
Thank you.  Everything is functionally working but it is not changing the icon color to yellow when the light is switched on.  Can you look at the css code and help me figure out why the icons are not changing to yellow?

I have modified my card language to be as follows:

  - title: Floorplan3
    panel: true
    cards:
      - type: 'custom:floorplan-card'
        full_height: true
        config: /local/floorplan/floorplan.yaml

_floorplan.yaml_
      name: Marks Floorplan
      image: /local/floorplan/floorplan.svg
      stylesheet: /local/floorplan/floorplan.css


      groups:
        - name: Lights
          entities:
             - light.caseta_r_wireless_in_wall_dimmer_2
             - light.caseta_r_wireless_in_wall_dimmer_3
             - light.caseta_r_wireless_in_wall_dimmer_4
             - light.caseta_r_wireless_in_wall_dimmer_5
             - light.driveway_lights
          states:
            - state: 'on'
              class: 'light-on'
            - state: 'off'
              class: 'light-off'
          action:
            domain: homeassistant
            service: toggle

      last_motion_entity: sensor.template_last_motion
      last_motion_class: last-motion
      defaults:
        hover_action: hover-info
      rules:
        - entity: light.caseta_r_wireless_in_wall_dimmer_5
          element: light.caseta_r_wireless_in_wall_dimmer_5
          tap_action: toggle
        - entity: light.caseta_r_wireless_in_wall_dimmer_4
          element: light.caseta_r_wireless_in_wall_dimmer_4
          tap_action: toggle
        - entity: light.caseta_r_wireless_in_wall_dimmer_3         
          element: light.caseta_r_wireless_in_wall_dimmer_3
          tap_action: toggle
        - entity: light.caseta_r_wireless_in_wall_dimmer_2
          element: light.caseta_r_wireless_in_wall_dimmer_2
          tap_action: toggle
        - entity: light.driveway_lights
          element: light.driveway_lights
          tap_action: toggle

_And floorplan.css_
/* SVG shapes */
svg, svg * {
  vector-effect: non-scaling-size !important;
  pointer-events: all !important;
}

/* SVG size */

#floorplan  {
  padding: 10px;
}

#floorplan > svg {
  max-width: 700px;
}

/* SVG elements */

svg * {
  vector-effect: non-scaling-stroke !important;
}

/* Hover over */

.floorplan-shape:hover,
g.floorplan-hover > :not(text):hover,
g.floorplan-click > :not(text):hover,
g.floorplan-long-click > :not(text):hover,
:not(text).floorplan-hover:hover,
:not(text).floorplan-click:hover, 
:not(text).floorplan-long-click:hover {
  stroke: #03A9F4 !important;
  stroke-width: 1px !important;
  stroke-opacity: 1 !important;
}

/* Animation */

.spinning {
  animation-name: spin;
  animation-duration: 5s;
  animation-iteration-count: infinite;
  animation-timing-function: linear;
  transform-origin: 50% 50%;
  transform-box: fill-box;
}

@keyframes spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

/* Binary sensors */

.binary-sensor-on {
  fill: #F9D27C !important;
}

.binary-sensor-off {
  fill: #7CB1F9 !important;
  transition: fill 5s ease;
}

/* Buttons */

.button-on rect {
  fill: #1ABA92 !important;
}

.button-off rect {
  fill: #d32f2f !important;
}

.button-on tspan,
.button-off tspan {
  fill: white !important;
}

/* Light */

.light-off {
  fill: #000000 !important;
}

.light-on {
  fill: #FFFF00 !important;
}

**Reply by vajonam** [reply]
Can you do an inspect using chrome on the floorplan, specifically on the bulb. on mine I have rect that I change the color on. 

![image](https://user-images.githubusercontent.com/152501/105505185-a51e2c80-5c96-11eb-98da-0d9b865ee213.png)

does it show the `class="light-off"` or something like mine above I use `button-off`, you can than manually change it see if actually changes.

**Reply by gillman99** [reply]
Vajonam, thanks for your help.  I am not seeing the class=light-off on inspect.  I've attached the inspect code below.  Not sure why it's not getting my class from the floorplan.yaml file.  It is creating 2 new classes which I believe are defined in the css above.


<img width="321" alt="Screen Shot 2021-01-22 at 8 56 19 AM" src="https://user-images.githubusercontent.com/76920134/105506468-b3b51580-5c8f-11eb-9375-9bc0899abca0.png">

**Reply by vajonam** [reply]
First off I have just been using this for a few days. 

```
  - entity: light.cresha_blah
    element: light.cresha_blah
    state_action:
      action: call-service
      service: floorplan.class_set
      service_data: "light-${entity.state}"
    tap_action:
      action: call-service
      service: homeassistant.toggle
```
Try that. 

This is how I toggle. I dont have a `groups:` section I have it all in the rules section. Since the docs are still being worked on I am not sure what the `groups:` config does..

**Comment by gillman99** [comment]
I will give this a try and let you know how it goes.
Thanks.

**Reply by vajonam** [reply]
I think the `groups:` section is hangover from older ha-floorplan I would use the format specified in 

https://experiencelovelace.github.io/ha-floorplan/docs/example-home/

**Reply by vajonam** [reply]
from the source you can see that only following sections are expected in the config.

https://github.com/ExperienceLovelace/ha-floorplan/blob/251a1eaa0af31d881a06317ff52679f97746821c/src/components/floorplan/lib/floorplan-config.ts#L13

**Reply by gillman99** [reply]
vajonam,
removing the groups section and using this in my floorplan.yaml file I can now see the class light-on and light-off.  But the icon does not change colors based on the fill command in the floorplan.css file.

        - entity: light.caseta_r_wireless_in_wall_dimmer_5
          element: light.caseta_r_wireless_in_wall_dimmer_5
          state_action:
            action: call-service
            service: floorplan.class_set
            service_data: "light-${entity.state}"
          tap_action:
            action: call-service
            service: homeassistant.toggle

I've tried the template in your last recommendation but that leaves me with an icon that's not the same size as the icon in my floorplan.svg file.  I haven't spent time trying to debug that.

In Inspect, I can see the class change from light-on to light-off when I click it but icon always stays black.  Any ideas?

**Reply by vajonam** [reply]
Not sure which icon you are changing, if its just an outline you will have to set the `stroke: #ff0000` as well as fill, My recommendation is that you can do all the test in the chrome inspector. you can right click on the bulb click inspect change the class or in the class edit the css on the right pane and see what you need to send in order to change it, and then modify the css accordingly.

**Reply by gillman99** [reply]
Sorry to keep bugging you with basic questions and I appreciate your help. I've attached a screenshot so you can see what I'm seeing.  I see the class light-off change to light-on and change to yellow when I click the light, but it stays dark.  Further up I see in element a fill that always stays RGB(0,0,0) but it's crossed out which i'm not sure what that means.  

I'm not able to manually edit any of the fields in inspect to change the color.  Not to say there's not a way I just don't know how.

<img width="936" alt="Screen Shot 2021-01-22 at 10 09 53 AM" src="https://user-images.githubusercontent.com/76920134/105515166-faa80880-5c99-11eb-9fa2-528ba572efb8.png">

**Reply by vajonam** [reply]
for whatever reason your css is "commented out" see the `/* */` tags. 

also have look at https://developers.google.com/web/tools/chrome-devtools/css/overrides#:~:text=The%20first%20thing%20to%20do,this%20concept%20is%20called%20Specificity.

**Reply by vajonam** [reply]
also maybe the floorplan-hover class is interfering as well?

**Comment by gillman99** [comment]
Thank you Vajonam.  I'll review the google doc and spend more time playing with this.  I'm sure I'll come back with more questions but hopefully they'll be more intelligent questions.
Thanks for your help.

**Comment by gillman99** [comment]
Vajonam,
I went back and used pretty much the exact same code and light icons as the example.  Everything is functional and the icon changes from clear to yellow when the light is turned on.  I'm attaching my new yaml, css and svg files.  You can see from the image below that the light icon I'm playing with (the clear one) is not sized or placed in the same place as floorplan.svg puts it.  I've been playing around with inspect and not sure why the icon is not coming in sized or placed correctly.  I appreciate your help earlier and thought you may have some ideas.

**floorplan.svg**
<img width="633" alt="Screen Shot 2021-01-22 at 5 38 46 PM" src="https://user-images.githubusercontent.com/76920134/105560160-b0924780-5cd8-11eb-80d2-aacb4b2fb59b.png">

**Screenshot of HA**
<img width="677" alt="Screen Shot 2021-01-22 at 5 39 22 PM" src="https://user-images.githubusercontent.com/76920134/105560193-c6077180-5cd8-11eb-894f-cf02ecf73cda.png">

Floorplan.yaml:
      defaults:
        hover_action: hover-info
        tap_action: more-info

      rules:
        - entity: light.caseta_r_wireless_in_wall_dimmer_5
          element: light.caseta_r_wireless_in_wall_dimmer_5
          state_action:
            action: call-service
            service: floorplan.image_set
            service_data: /local/floorplan/light_${entity.state}.svg
          tap_action:
            action: toggle

        - entity: light.caseta_r_wireless_in_wall_dimmer_5
          element: light.caseta_r_wireless_in_wall_dimmer_5.button
          state_action:
            action: call-service
            service: floorplan.class_set
            service_data: 'button-${entity.state}'
          tap_action:
            action: call-service
            service: homeassistant.toggle

[floorplan.svg.zip](https://github.com/ExperienceLovelace/ha-floorplan/files/5859189/floorplan.svg.zip)
[floorplan.css.zip](https://github.com/ExperienceLovelace/ha-floorplan/files/5859190/floorplan.css.zip)
[light_off.svg.zip](https://github.com/ExperienceLovelace/ha-floorplan/files/5859194/light_off.svg.zip)
[light_on.svg.zip](https://github.com/ExperienceLovelace/ha-floorplan/files/5859195/light_on.svg.zip)

**Comment by vajonam** [comment]
Remove the image set call and just change colours. If you want to make it
hidden set the opacity or visibility. I am not sure which order the rules
are fired. Not sure if they can be predicted. You assuming the order of the
rules right?

**Comment by gillman99** [comment]
What is the best way to change the color?  Is it done in yaml or css?  Can you send me command?

**Comment by vajonam** [comment]
I change the colour by setting the class but I think you can set style as
well but set class is a clean way to do it

Basically what your saying in the yaml that based on the state change the
class to something that includes the entity state like foo-(on) or
foo-(off)

**Reply by vajonam** [reply]
Here is what I suggest. That you put all the lights bulbs etc on on SVG. Then you assign the object names correct into the SVG. Then in the css you create 2 classes on for `-on` state and `-off` state. The off state will have the fill black and the on state will have fill as yellow. No need to switch between svg.  I try out your svgs and let you know

**Comment by vajonam** [comment, marked as answer]
### home assistant Lovelace yaml

```
  - type: 'custom:floorplan-card'
    title: Main Floor
    full_height: true
    config: /local/floorplan/floorplan.yaml?v=2
```


### yaml

```
full_height: false
image: /local/floorplan/floorplan.svg?v=2
stylesheet: /local/floorplan/home.css
# log_level: debug
console_log_level: debug
defaults:
  hover_action: more-info
  state_action:
    action: call-service
    service: floorplan.class_set
    service_data: "button-${entity.state}"
  tap_action:
    action: call-service
    service: homeassistant.toggle
rules:
  - entity: light.shower_lamp
    element: light.caseta_r_wireless_in_wall_dimmer_3
    state_action:
      action: call-service
      service: floorplan.class_set
      service_data: "button-${entity.state}"
```

### svg 
just use the floorplan.svg

### css

```
.button-on  {
  fill: rgb(245 255 0)  !important;
}

.button-off {
  fill: #494949 !important;
  /* transition: fill 1.5s ease; */
}


```

### result

![image](https://user-images.githubusercontent.com/152501/105562691-3cac6b00-5ce9-11eb-9dc7-d1d7c622590c.png)

![image](https://user-images.githubusercontent.com/152501/105562704-433ae280-5ce9-11eb-92c0-97539b400bba.png)

**Reply by vajonam** [reply]
hope this helps

**Comment by gillman99** [comment]
Thanks.  Giving it a shot now.  Let you know how it goes.
Mark

**Comment by gillman99** [comment]
Vajonam,
This is working perfectly now!  Thank you so much for your help.
Mark

**Comment by vajonam** [comment]
No worries. Remember chrome inspector is your friend.
